### PR TITLE
SQL: Fix esType for DATETIME/DATE and INTERVALS

### PR DESCRIPTION
**Pull request by matriv**
Since introduction of data types that don't have a corresponding type
in ES the `esType` is error-prone when used for `unmappedType()` calls.
Moreover since the renaming of `DATE` to `DATETIME` and the introduction
of an actual date-only `DATE` the `esType` would return `datetime` which
is not a valid type for ES mapping.

Fixes: #38051

### CODE REVIEW

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 1]
```diff
@@ -661,17 +661,17 @@ public void testTopHitsAggregationWithTwoArgs() {
 
         }
         {
-            PhysicalPlan p = optimizeAndPlan("SELECT LAST(keyword, int) FROM test");
+            PhysicalPlan p = optimizeAndPlan("SELECT LAST(date, int) FROM test");
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: Was this change required by the types change in this PR or you just wanted some diversity in FIRST/LAST tests?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java [hunk 2]
```diff
@@ -126,9 +127,9 @@
 
 
     /**
-     * Elasticsearch type name
+     * Lowercase type name
      */
-    public final String esType;
+    public final String esSQLType;
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: I may be wrong, but why not `esSqlType` name?

Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: sure

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 1 above)
Comment (matriv) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: I wanted to test the `unmapped_type` https://github.com/elastic/elasticsearch/pull/38179/files#diff-aef2b0ce456b8fdd5cc09d6cfd55f0c2R674 which before the fix was `datetime`

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java [hunk 3]
```diff
@@ -126,9 +127,9 @@
 
 
     /**
-     * Elasticsearch type name
+     * Lowercase type name
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: I am not sure "Lowercase" is the right description for this field, I mean it is a true statement, but doesn't help much with the purpose of the field.

**Review by astefan — APPROVED** [review]
LGTM. Left just few minor comments.

**Review by costin — CHANGES_REQUESTED** [review]
I don't see why a new property was introduced (esSQLType - by the way the naming is incorrect since we use Sql everywhere else) or what does it even mean - the type of esSQL?
If it's the Elasticsearch type then that's what `esType` is for. if it's the sql type it's `sqlType`.

If it's the type name itself why not use dt.name? If it's a lowercase vs uppercase (again arguable since it's es vs sql) a typeName() method might be more appropriate.

Further more the style of the code changes is confusing - the field access (.esType) was moved to method invocation (esType()) but not in all places.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java (hunk 3 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: +1. 
The purpose of a comment is to explain the non-obvious:
```
/** an integer */ vs /** array index */
public int i;
```

**Comment by costin** [comment]
> So the idea is to use the lowercase name to everywhere it has to do with error messages or returning the column data type to the client as we used to, and use the new method `.esType()` that throws the exception and treats the `DATETIME`->`date` correctly wherever we use it for calls to ES APIs, like in the `FieldHitExtractor` or the `TopHitsAgg`.

Being a enum, `name()` already exists for  the upper case variant and assuming we use it a lot, can define `typeName` for error messages.
My point still stands that the style should be consistent (use either the field or the method not both).
I also find it weird to throw an exception when calling esType instead of returning null.
It's defensive programming that I'm not a fan of plus the caller is already aware (otherwise one would get an exception).

Let's discuss this more next week.

**Comment by matriv** [comment]
So, for error messages, return column types, etc.. we need the name of the enum in lowercase.
And we want to have there `date`, `datetime`, and the interval types.

BUT for calls to `unmappedType` that we use when we do field extraction we need something else, since our `DATETIME` maps to ES `date`, so if we use the same method/field whatever as above that just lowercase we will call `unmapedType` with `datetime` for example which is not a valid type for ES.

Moreover if we try to make a call to `unmappedType`, which happens during field extraction with a DataType that doesn't exist in ES (like `DATE` or `INTERVALxxx` then returning null is not safe, imho, as null is a valid value for `unmappedType` method. We should never try to ask from ES to extract a field with type `INTERVALxxx`, therefore I chose the exception.

**Comment by matriv** [comment]
@costin Please check again.
Removed completely the check for ES not mapped types (DATE & Intervals) as in the places we require the `esType` we have a `FieldAttribute` with the proper `EsField` using the `fromTypeName`. My idea for the exception was to be bug-proof for the future, but I totally understand your point about "defensive" programming.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java [hunk 4]
```diff
@@ -176,7 +181,8 @@
 
     DataType(SQLType sqlType, int size, int defaultPrecision, int displaySize, boolean isInteger,
             boolean isRational, boolean defaultDocValues) {
-        this.esType = name().toLowerCase(Locale.ROOT);
+        this.typeName = name().toLowerCase(Locale.ROOT);
+        this.esType = name().equals("DATETIME") ? "date" : typeName;
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: I think a better way would be to ask this parameter for the constructor and add it for each declaration. 
So `DATETIME` would be `"date"`, `INTERVAL_`... `null` while `NULL` would be `"null"`. Potentially add an overloaded constructor to specify esType name which by default would be the enum name lowercase but with the possibility of being overwritten.

**Review by costin — COMMENTED** [review]
Looks good however there's still the esType initialization that could be sorted out.

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java [hunk 5]
```diff
@@ -23,37 +23,37 @@
...
-    NESTED(                JDBCType.STRUCT,    -1,                0,                 0,  false, false, false),
-    BINARY(                JDBCType.VARBINARY, -1,                Integer.MAX_VALUE, 0,  false, false, false),
-    DATE(                  JDBCType.DATE,      Long.BYTES,        10,                10, false, false, true),
+    SCALED_FLOAT(  "scaled_float",   JDBCType.FLOAT,     Double.BYTES,      19,                25, false, true,  true),
+    KEYWORD(       "keyword",        JDBCType.VARCHAR,   Integer.MAX_VALUE, 256,               0,  false, false, true),
+    TEXT(          "text",           JDBCType.VARCHAR,   Integer.MAX_VALUE, Integer.MAX_VALUE, 0,  false, false, false),
+    OBJECT(        "object",         JDBCType.STRUCT,    -1,                0,                 0,  false, false, false),
+    NESTED(        "nested",         JDBCType.STRUCT,    -1,                0,                 0,  false, false, false),
+    BINARY(        "binary",         JDBCType.VARBINARY, -1,                Integer.MAX_VALUE, 0,  false, false, false),
+    DATE(                            JDBCType.DATE,      Long.BYTES,        10,                10, false, false, true),
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: Shouldn't this be "date" as well?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java [hunk 6]
```diff
@@ -23,37 +23,37 @@
...
-    //                     jdbc type,          size,              defPrecision,dispSize, int,   rat,   docvals
-    NULL(                  JDBCType.NULL,      0,                 0,                 0,  false, false, false),
-    UNSUPPORTED(           JDBCType.OTHER,     0,                 0,                 0,  false, false, false),
-    BOOLEAN(               JDBCType.BOOLEAN,   1,                 1,                 1,  false, false, false),
-    BYTE(                  JDBCType.TINYINT,   Byte.BYTES,        3,                 5,  true,  false, true),
-    SHORT(                 JDBCType.SMALLINT,  Short.BYTES,       5,                 6,  true,  false, true),
-    INTEGER(               JDBCType.INTEGER,   Integer.BYTES,     10,                11, true,  false, true),
-    LONG(                  JDBCType.BIGINT,    Long.BYTES,        19,                20, true,  false, true),
+    //             esType            jdbc type,          size,              defPrecision,dispSize, int,   rat,   docvals
+    NULL(          "null",           JDBCType.NULL,      0,                 0,                 0,  false, false, false),
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: I would have gone the other way - use the lower-case enum way as a default type and just override this convention with null or "date". Just a matter of preference really.

**Review by costin — APPROVED** [review]

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java (hunk 6 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: it's more elegant yes, but I choose the more "explicit" way to show clearly what the corresponding esType is.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java (hunk 5 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataType.java]: But `DATE` doesn't have a corresponding esType, same as all INTERVALS.
